[zyonwee] iP

zyonwee opened the pull request:
# Bob: Your Personal Task Manager 🚀
![alt text](https://github.com/zyonwee/iP/blob/master/src/main/resources/images/user.png?raw=true)  

Bob is a lightweight, text-based task manager designed to help you organize your tasks efficiently. Inspired by the quote:

> “Your mind is for having ideas, not holding them.” – [David Allen](https://gettingthingsdone.com/)

Bob frees your mind from the burden of remembering tasks. It’s:

*   Text-based: Simple and intuitive.
*   Easy to learn: No steep learning curve.
*   **FAST SUPER FAST** to use: Get things done in seconds.

## How to Get Started

1.  Download Bob from [here](https://github.com/zyonwee/ip/releases/download/v0.2/bob.jar).
2.  Double-click the executable file.
3.  Add your tasks using simple commands.
4.  Let Bob manage your tasks for you 😉.

And the best part? It’s FREE! 🎉

## Features

-   Add tasks with deadlines, events, or todos.
-   Mark tasks as done or undone.
-   Delete tasks you no longer need.
-   Save and load tasks automatically.
-   ~~Sync tasks across devices~~ (coming soon!).

## Example Usage

Here’s how you can use Bob:

```java
// Add a todo task
todo Buy groceries

// Add a deadline task
deadline Submit report /by 2023-12-31

// Add an event task
event Team meeting /from 2023-10-01 /to 2023-10-02

// Mark task No. 1 as done
mark 1

// Unmark task No. 1
unmark 1

// Delete task No. 1
delete 1

// Find a task called 'Cry'
find cry

// Lists all tasks
list  

// Exit application
exit
```
# Acknowledgements

This project incorporates code and/or ideas from the following sources:

*   **Code Reuse:**  The `DialogBox` `MainWindow` classes is adapted from work by Wang Haitao iP. 

*   **AI Assistance:** Gemini and DeepSeek were used for generating initial drafts of Javadoc documentation.

![alt text](https://github.com/zyonwee/iP/blob/master/docs/Ui.png?raw=true)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

zyonwee committed:
Merge branch 'branch-Level-10'

Add Varargs

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

zyonwee committed:
rename file and class

zyonwee committed:
remove duke in README.md

zyonwee committed:
Add greeting

zyonwee committed:
Add exit message

zyonwee committed:
Added echoMsg method

zyonwee committed:
Added exit trigger

zyonwee committed:
Added handleCommand method

zyonwee committed:
Added List Commands

zyonwee committed:
Added Task class

zyonwee committed:
Added mark task

zyonwee committed:
Added unmark command

zyonwee committed:
Added ToDo Class

zyonwee committed:
Added Deadline Class

zyonwee committed:
Added Event Class

zyonwee committed:
Added input.txt, EXPECTED.TXT, Test Cases

zyonwee committed:
Added Count and Details of Successful Test Cases

zyonwee committed:
Added BobException Class

zyonwee committed:
Added Delete Command and Test Cases

zyonwee committed:
Added CommandType Enum

zyonwee committed:
MainApp.java: Remove unuse import

zyonwee committed:
README.md: Update line space

zyonwee committed:
Launcher.java: Remove unuse import

zyonwee committed:
AddDeadlineCommand.java: Update comment

KarKuen commented on src/main/java/Bob.java:
```diff
@@ -0,0 +1,227 @@
+    static String greetMsg = "Hello! I'm " + fullName + "\n" +
+            "What can I do for you?\n" ; // Greeting Message
+    static String exitMsg =  "Bye. Please don't trouble me again!\n"; // Exit Message
+    static boolean listenMode = true; // Check if Bot should continue to listen for commands
```
Perhaps a more Boolean sounding variable name will make this variable more obviously Boolean, maybe "isListening" ?

KarKuen commented on src/main/java/Bob.java:
```diff
@@ -0,0 +1,227 @@
+    /**
+     * Handles the user input
+     */
+    public static void handleCommand() {
```
```suggestion
    public static void handleCommands() {
```

KarKuen commented on src/main/java/Bob.java:
```diff
@@ -0,0 +1,227 @@
+     * Adds Todo task to TaskList
+     * @param desc of ToDo Task
+     */
+    public static void addToDo(String desc) {
```
Perhaps you can include the word "Task" into the method name to make the function's purpose more obvious, maybe "addToDoTask" ?

KarKuen commented on src/main/java/CommandType.java:
```diff
@@ -0,0 +1,30 @@
+        this.commandString = commandString;
+    }
+
+    public static CommandType fromString(String command) {
```
Perhaps you can rename this method to describe its function ?

KarKuen reviewed (commented):
Overall nicely written code. Nice Job Zyon :)

arshinsikka commented on src/main/java/Bob.java:
```diff
@@ -0,0 +1,227 @@
+
+    static String fullName = "Bob the Princess"; // Name of Bot
+    static String greetMsg = "Hello! I'm " + fullName + "\n" +
+            "What can I do for you?\n" ; // Greeting Message
```
The string concatenation here is misaligned. Consider aligning both lines properly or using String.format() to improve readability.

arshinsikka commented on src/main/java/Bob.java:
```diff
@@ -0,0 +1,227 @@
+                            String eventTo = cleanString.substring(toIndex + 3);
+                            addEvent(description, eventFrom, eventTo);
+                        } catch (Exception e) {
+                            System.out.println("Did you forget your '/from' and '/to' command? So un-attentive!\n" +
```
Using System.out.println() for error messages makes debugging harder. Would using Logger instead improve maintainability and debugging?"

arshinsikka commented on src/main/java/Bob.java:
```diff
@@ -0,0 +1,227 @@
+                    case DELETE:
+                        // Handle delete command (similar to MARK)
+                        try {
+                            int taskNumber = Integer.parseInt(command.substring(7));
```
The try-catch blocks for parsing task numbers are duplicated across multiple commands. Would extracting a helper method like parseTaskNumber() improve reusability and reduce redundancy?"

arshinsikka reviewed (commented):
Overall, looks good to me, it was very tough to find errors.

zyonwee committed:
Add Save Task

zyonwee committed:
Merge branch 'branch-Level-7'

zyonwee committed:
Bob class: rename Boolean variables

zyonwee committed:
Add Date Utils Class

zyonwee committed:
Merge branch 'branch-Level-8'

zyonwee committed:
Update MORE OOP

zyonwee committed:
Update package structure

zyonwee committed:
Merge branch 'add-gradle-support'

zyonwee committed:
Add JUnit Tests

zyonwee committed:
build.gradle: Create a JAR File

zyonwee committed:
Add Find Command

zyonwee committed:
Add Java Doc

zyonwee committed:
Add Java Doc

zyonwee committed:
Update Coding Standard

zyonwee committed:
Merge branch 'branch-A-CodingStandard'

zyonwee committed:
Merge branch 'branch-A-JavaDoc'

zyonwee committed:
Add Check Style

zyonwee committed:
README.md: Update Bob

zyonwee committed:
Merge branch 'Use-GFMD'

zyonwee committed:
Add GUI

zyonwee committed:
Merge branch 'branch-Level-10'

zyonwee committed:
Merge branch 'branch-A-Varargs'

zyonwee committed:
Update Java Doc

zyonwee committed:
Revert "Update Java Doc"

This reverts commit cb7171cf255d682d5dc48a98f7172773399284fb, reversing
changes made to 7bf490c2ca380a6912c3b8401bcd88abc1bdf749.

zyonwee committed:
Merge branch 'master' of https://github.com/zyonwee/ip

zyonwee committed:
Add Java assertions to Storage class for data integrity

The `Storage` class loads and saves tasks to a file.  It currently lacks
explicit checks to ensure data integrity.

This can lead to unexpected behavior or errors if, for example, the file
path is invalid, corrupted, or the task data is inconsistent.

Let's add Java assertions to the `Storage` class to validate inputs and
internal state.  Specifically, let's add assertions to:

* The constructor to check for null or empty file paths.
* The `load()` method to ensure parsed tasks are not null.
* The `save()` method to verify that the task list and individual tasks
  are not null.

Assertions provide runtime checks during development and testing, helping
to catch potential programming errors related to data integrity early.
Using assertions in these specific locations targets the most critical
points where data corruption or invalid inputs could occur.  Assertions
are disabled by default in production, so they don't impact performance
in the final product.  They serve as a valuable debugging tool during
development.

zyonwee committed:
Merge branch 'branch-A-Assertions'

zyonwee committed:
Refactor CommandType.fromString() using Java Streams

The `CommandType.fromString()` method currently uses a loop to find the
matching command type.

This can be made more concise and readable using Java Streams.

Let's refactor the `fromString()` method to use Java Streams.  This
replaces the explicit loop with a functional approach, making the code
easier to understand and maintain.  The use of Streams also improves
the handling of cases where the input string is empty or no matching
command type is found, using the `Optional` type effectively.

This change improves the code quality and readability of the
`CommandType` enum.

zyonwee committed:
Merge branch 'branch-A-CodeQuality'

zyonwee committed:
DialogBox.fxml : Improve dialog box styling

This commit updates the dialog box appearance with:

*   Darker and more distinct background colors for user and Bob messages.
*   A light grey dotted border for better visual separation.
*   Ensures consistent styling across all dialog boxes.

These changes improve the overall user experience and readability.

zyonwee committed:
DialogBox.java: Generate Javadoc with AI assistance

This commit uses an AI tool to generate the initial Javadoc for
the DialogBox class.

The generated Javadoc has been reviewed and edited to ensure accuracy
and completeness.  Further manual review and updates may be required.

zyonwee committed:
BobException.java: Implement sassy personality for Bob exceptions

This commit adds a touch of sass to Bob's exception messages.
Error messages now include witty prefixes and suffixes,
reflecting Bob's playful and slightly sarcastic personality.

This change enhances the user experience by adding a bit of humor
to error situations. The sassy messages are intended to be
light-hearted and engaging, while still conveying important information
about the error.

zyonwee committed:
README.md: Update user guide for [mention specific changes or features]

This commit updates the user guide to reflect the following changes:

*   [List specific changes, e.g., Updated instructions for adding tasks.]
*   [Added a new section on using the search functionality.]
*   [Improved examples for deadline and event creation.]

These changes improve the user experience by providing up-to-date info.

zyonwee committed:
Revert "README.md: Update user guide for [mention specific changes or features]"

This reverts commit a208414f85a345c30affb035bf51b0df72a455db.

zyonwee committed:
README.md: Update user guide

This commit updates the acknowledgements section of the user guide
to include:

*   Acknowledgement of code reuse from Wang Haitao iP.
*   Acknowledgement of AI assistance from Gemini and DeepSeek
for Javadoc generation and code suggestions.

These changes ensure proper attribution and transparency in the project.

zyonwee committed:
Ui.png: Add Ui.png screenshot to documentation

This commit adds Ui.png, a screenshot of the user interface,
to the User Guide.

This image helps users visualize the application's interface
and understand its layout.

zyonwee committed:
README.md: Update README with complete command reference

This commit updates the README file to include a comprehensive list of
all available commands, along with their descriptions and usage
examples.

The added command reference provides users with a complete guide to
interacting with the application.  It includes details on how to add
different types of tasks (todo, deadline, event), manage tasks (mark,
unmark, delete, list), and use other helpful commands (help, bye).

zyonwee committed:
README.md: Add JAR download instructions and link

This commit updates the README file to include instructions on how to
download and use the pre-built JAR file for the application.

The README now includes a direct link to the JAR file and provides
basic usage instructions for running the application from the JAR.
This makes it easier for users to get started with the application
without needing to build it from source.

zyonwee committed:
README.md: Updated bob.jar link

zyonwee committed:
DateUtilsTest.java: Add tests for DateUtils

zyonwee committed:
Storage.java: Add Error Handling

zyonwee committed:
Merge branch 'branch-UserGuide'

zyonwee committed:
Merge branch 'branch-A-ErrorHandling'

zyonwee committed:
Storage.java: Fix File Error

zyonwee committed:
Update README.md

zyonwee committed:
Update README.md
